packages/serverpod_test/lib/src/function_call_wrappers.dart

```diff
@@ -0,0 +1,76 @@
+import 'dart:async';
+
+import 'package:serverpod/serverpod.dart';
+
+/// Test tools helper to not leak exceptions from awaitable functions.
+/// Used by the generated code.
+Future<T> callAwaitableFunctionAndHandleExceptions<T>(
+  Future<T> Function() call,
+) async {
+  try {
+    return await call();
+  } catch (e) {
+    throw _getException(e);
+  }
+}
+
+/// Test tools helper to not leak exceptions from functions that return streams.
+/// Used by the generated code.
+Future<void> callStreamFunctionAndHandleExceptions<T>(
+  Future<Stream<T>> Function() call,
+  StreamController<T> streamController,
+) async {
+  late Stream<T> stream;
+  try {
+    stream = await call();
+  } catch (e) {
+    streamController.addError(_getException(e));
+    return;
+  }
+
+  var subscription = stream.listen((data) {
+    streamController.add(data);
+  }, onError: (e) {
+    streamController.addError(_getException(e));
+  }, onDone: () {
+    streamController.close();
+  });
+
+  streamController.onCancel = () {
+    subscription.cancel();
+  };
+}
+
+/// The user was not authenticated.
+class UnauthenticatedEndpointCallTestException implements Exception {
+  /// Creates a new UnauthenticatedEndpointCallTestException.
+  UnauthenticatedEndpointCallTestException();
+}
+
+/// The authentication key provided did not have sufficient access.
+class InsufficientEndpointAccessTestException implements Exception {
+  /// Creates a new InsufficientEndpointAccessTestException.
+  InsufficientEndpointAccessTestException();
+}
+
+dynamic _getException(dynamic e) {
+  switch (e) {
+    case NotAuthorizedException():
+      return switch (e.authenticationFailedResult.reason) {
+        AuthenticationFailureReason.unauthenticated =>
+          UnauthenticatedEndpointCallTestException(),
+        AuthenticationFailureReason.insufficientAccess =>
+          InsufficientEndpointAccessTestException(),
```

Comment:
Suggestion: As discussed offline, it would be nice to align the clientside exceptions with the exceptions in the tests.

Reply:
@Isakdl I think we ended up at the conclusion that the client exceptions probably should be named something that is not tied 1:1 to HTTP response statuses. So the question is, are these names okay or do we still want to change them?

Reply:
Yes that is fair, but I do think they should map 1:1 with the names and exceptions we use in the client. Even if that means renaming / extending new exceptions for the client.

Reply:
[b88616b](https://github.com/serverpod/serverpod/pull/2713/commits/b88616b98a5f3eff87e430f097f3deb2d82afe48)